Feature/Bookmarks: add & edit bookmarks

felix-schwarz opened the pull request:
## Changes
- New EditBookmarkViewController implementation:
   - dynamic animation of all fields
   - automatic input focus change (+ scrolling)
   - progress hud indicating actions taking place / blocking access to fields during operations
   - shake animation for wrong credentials
   - indication of authentication method for token based auth methods
- New ProgressHUDViewController that can be used to block access and provide full screen status feedback during operations
- New DispatchQueueTools: "OnMainQueue" as a replacement for "DispatchQueue.main.async"
- StaticTableView updates
   - new property attached: indicates if a row or section is currently part of a static table view hierarchy
   - new property index: index of the section / row in the respective part of the static table view
   - new property indexPath: indexPath of the row in the static table view
   - new property needsLiveUpdates : indicates if the view controller is currently shown and live updates of the table view are desirable
   - added animation support for adding rows to sections
   - added support for inserting and removing rows from sections (incl. animation)
   - added enabled and updateViewAppearance support to rows
   - added new "label" row
- New UIView+Animation extension to provide simple animations for all UIViews (right now only shakeHorizontally)
- New OCConnectionIssue extension that can prepare an issue (incl. groups) for display
- Updated CoreManager to take advantage of the latest SDK changes (and fix a bug)
- Fixed project settings
- Removed debris in ownCloud subdirectory from previous merge
- Add support for password managers (tested with 1Password), closing #30 
- Add support for OCClassSettings (url-editable, default-url)
- Add error handling to ClientRootViewController with special case handling for wrong credentials (now offers editing them)
- Add support for disabled text/secure text fields in StaticTableView
- "Save" button in edit mode now works correctly
- Turned off auto-correct for username and password text fields
- Updated ios-sdk (**IMPORTANT: YOU NEED TO DELETE AND RECREATE YOUR BOOKMARKS FOR THINGS TO WORK CORRECTLY**), closing #35

## Before you review
Please make sure to delete all your existing bookmarks before proceeding. A syntax error in a SQLite trigger will otherwise cause weird issues. This could have been fixed with a migration as well, but I believe time is better spent elsewhere.

____

BUGS & IMPROVEMENTS

- [X] (1) Empty URL https://github.com/owncloud/ios-app/pull/38#issuecomment-387765168 [FIXED]
- [x] (2) Authorization failed dialog https://github.com/owncloud/ios-app/pull/38#issuecomment-387778175 [SDK issue]
- [X] (3) Default setting to make the URL editable (lock icon) https://github.com/owncloud/ios-app/pull/38#issuecomment-387781936 [FIXED]
- [ ] (4) Detection of auth methods in Edit mode https://github.com/owncloud/ios-app/pull/38#issuecomment-388011240

felix-schwarz commented:
Here's how OAuth2 bookmarks look when editing. The username of the authenticated user is now also shown.

<img width="375" alt="png-bild 3" src="https://user-images.githubusercontent.com/639669/39786262-13ec80e0-5320-11e8-813c-2566c26399f5.png">

Password managers like 1Password also are supported now:

<img width="375" alt="png-bild 2" src="https://user-images.githubusercontent.com/639669/39786271-24c44240-5320-11e8-87ac-1933938be4d7.png">

felix-schwarz commented on ownCloud/Bookmarks/BookmarkViewController.swift:
```diff
@@ -20,267 +20,747 @@ import UIKit
+		var userName : String? = nil
+		var password : String? = nil
+
+		if let authMethodIdentifier = bookmark.authenticationMethodIdentifier {
```
Good catch! Done.

felix-schwarz commented on ownCloud/UIKit Extensions/UIView+Animation.swift:
```diff
@@ -0,0 +1,32 @@
+
+		animation.timingFunction = CAMediaTimingFunction(name: kCAMediaTimingFunctionLinear)
+		animation.duration = duration
+		// animation.values = [ -amplitude, amplitude, -(amplitude*2/3), (amplitude*2/3), -(amplitude*1/3), (amplitude*1/3), 0 ]
```
Another good catch! Done.

felix-schwarz commented on ownCloud/Client/ClientRootViewController.swift:
```diff
@@ -114,3 +123,62 @@ class ClientRootViewController: UITabBarController {
+							let editBookmark = self.bookmark
+
+							self.closeClient(completion: {
+								if presentingViewController != nil {
```
And another good catch! Done.

jesmrec commented:
(1)

Steps:

1. Set an empty URL and click on `Continue`

Current: Server URL field is set with `http:///`
Expected: Server URL field is empty, so no input was typed there

jesmrec commented:
As discussed before, i would set the username as non-editable in edit mode. Changing the username is creating a new bookmark. This would work as shortcut for "delete & create" but really, the action performed is not an `Edit`

jesmrec commented:
### (2)

Question:

This check:

![screen shot 2018-05-09 at 17 24 53](https://user-images.githubusercontent.com/14894746/39823733-f68e0e40-53ad-11e8-9758-5bc88b7322b0.png)

is currently performed at the time the bookmark is selected from the list and then, displayed over the file list. If credentials change during the session, is shown `Contents from cache` but not the dialog to let the user know his/her credentials are different. Is this one an expected behaviour? If not, would be a nice to have

jesmrec commented:
### (3)

Setting in 

https://github.com/owncloud/ios-app/blob/de9a8ae3abc5b885a5d9b2aff6e914d7992a30c0/ownCloud/Bookmarks/BookmarkViewController.swift#L706

`.bookmarkURLEditable : true`

the lock icon at the right of the URL is displayed. It should be displayed when the property is set to ´false´, so the user can not edit.

![gif1](https://user-images.githubusercontent.com/14894746/39824486-ba173728-53af-11e8-8298-29fac4c82342.gif)

jesmrec commented:
### (4)

Steps:

1. Add a basic auth bookmark
2. In server, enable OAuth2
3. Go to Edit bookmark

Current: Credentials and URL did not change, so the user will not perform any changes in `Edit` and then `Save`.  The connection will use basic auth credentials instead of OAuth2 ones.

Expected: URL should re-check available auth methods to get OAuth2 in case it is available in server side.

felix-schwarz commented:
> ### (1)
> 
> Steps:
> 
> 1. Set an empty URL and click on `Continue`
> 
> Current: Server URL field is set with `http:///`
> Expected: Server URL field is empty, so no input was typed there

Thanks! Fixed in latest commit.

> As discussed before, i would set the username as non-editable in edit mode. Changing the username is creating a new bookmark. This would work as shortcut for "delete & create" but really, the action performed is not an `Edit`

Thanks for the reminder. Made that change in the latest commit.

> ### (2)
> 
> Question:
> 
> This check:
> 
> ![screen shot 2018-05-09 at 17 24 53](https://user-images.githubusercontent.com/14894746/39823733-f68e0e40-53ad-11e8-9758-5bc88b7322b0.png)
> 
> is currently performed at the time the bookmark is selected from the list and then, displayed over the file list. If credentials change during the session, is shown `Contents from cache` but not the dialog to let the user know his/her credentials are different. Is this one an expected behaviour? If not, would be a nice to have

Looking into this I noticed that `demo.owncloud.org` does return an `500 Internal Server Error` (Request ID: 7A07F71E-E475-4780-8A4B-9167C02E66D5) for a standard PROPFIND, depth=1 on a folder, if the login credentials are no longer valid. 

The file list implementation will show that error whenever it receives it from the SDK. However, right now, the SDK only handles 404 errors for file lists. So this is something to address in the SDK.

> ### (3)
> 
> Setting in 
> 
> https://github.com/owncloud/ios-app/blob/de9a8ae3abc5b885a5d9b2aff6e914d7992a30c0/ownCloud/Bookmarks/BookmarkViewController.swift#L706
> 
> `.bookmarkURLEditable : true`
> 
> the lock icon at the right of the URL is displayed. It should be displayed when the property is set to ´false´, so the user can not edit.

Good catch! Fixed in the latest commit as well.

jesmrec commented:
(1) and (3) fixed
(2) will be addressed to SDK

felix-schwarz commented:
> ### (4)
> 
> Steps:
> 
> 1. Add a basic auth bookmark
> 2. In server, enable OAuth2
> 3. Go to Edit bookmark
> 
> Current: Credentials and URL did not change, so the user will not perform any changes in `Edit` and then `Save`.  The connection will use basic auth credentials instead of OAuth2 ones.
> 
> Expected: URL should re-check available auth methods to get OAuth2 in case it is available in server side.

You're raising a good point here.

I have a few thoughts on this:
- as a user I feel I should be able to make a change to f.ex. the name without being forced to upgrade the authentication mechanism.

- I can see the clear security win for users to upgrade to a better mechanism if one is available, so the check could be made when hitting "Save", but only lead to a suggestion - and not be forced upon the user.

- Users who don't edit their bookmarks wouldn't benefit from this, however, so… maybe we should periodically check available authentication methods in general and tell the user if something better is available.

What do you think?
